### DISCUSSION TITLE
Log both loss and metrics on both train  and eval dataset into a csv file

### DISCUSSION THREAD

**Discussion by anais2390** [Project-MONAI/tutorials#2023] [Q&A]
Hello, 
I am using monai librairy for segmentation and I've been looking for the cleanest way to log at each epoch:
-  mean **loss** computation on **train** dataset
-  mean **metrics** computations on **train** dataset
-  mean **loss** computation on **eval** dataset
-  mean **metrics** computations on **eval** dataset
- **learning rate**

I am using engine **SupervisedTrainer** and **SupervisedEvaluator** and the trickest part is that the output_transforms are different for loss computation (no postprocessing) and metrics computation (activation + asdiscrete).
Also I coudn't manage to get  the loss values into the csv.

A typical csv file sould contain the following columns (if we use the MeanDice metric for exemple)

Epoch  |   MeanDice_train  | MeanDice_eval  | Loss_train |  Loss_eval | 

Thank you so much for your help.

Anaïs

**Comment by ericspod** [comment]
Hi @anais2390 you can use the [MetricLogger](https://github.com/Project-MONAI/MONAI/blob/b5bc69d0306c16cba61de58a403190689bb192b3/monai/handlers/metric_logger.py#L41) class to capture all the metric and loss data during training and evaluation, probably with two separate instances. These will capture the values which can then be used to fill in a csv file afterward. You can also use event handlers to trigger at the end of an epoch to record the learning rate to a data structure somewhere, something like this:

```python
opt = SomeOptimizer()
epoch_lrs=[]
...

@trainer.on(Events.EPOCH_COMPLETED)
def get_lr(engine):
    epoch_lrs.append(opt.lr)
```
